CentOS 6 workaround for 7.6.0 (aeron, libmemcached, boost)

Michael-Gardner opened the pull request:
Signed-off-by: Michael Gardner <michael.gardner@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
centos 6 workaround for 7.6.0

Signed-off-by: Michael Gardner <michael.gardner@lexisnexisrisk.com>

Centos-6-workaround adding libboost-regex.so, build and import

Signed-off-by: Michael Gardner <michael.gardner@lexisnexisrisk.com>

HPCCSmoketest commented:
Process of PR-12959, label: centos-6-workaround-7.6.0 starts now.
The reason of this test is: Did not build yet.
Commit ID: 7b791c94c00158b90a09bb27e0a7101343f6d251
Estimated completion time is ~0.5 hour(s)

Michael-Gardner commented:
@xwang2713 Please take a look, not sure if we want to use this but I figured out a way to deal with centos 6 builds (at least temporarily).

Downloaded source for boost from here and unpacked into my home directory https://www.boost.org/users/download/
Extra stuff I added to the command line was this `cmake ../HPCC-Platform/ -DUSE_BOOST_REGEX=OFF -DMEMCACHED_USE_EXTERNAL_LIBRARY=ON -DBOOST_REGEX_INCLUDE_DIR=/home/michael/boost_1_71_0`

- added -fpermissive to libmemcached build we only ever use on CentOS 6.
- added linker flag -lrt when we go to include the aeron code
- worked around a bug with boost regex by using std library instead (USE_BOOST_REGEX=OFF)
- worked around some code requiring boost ptree templates by downloading the header files on my build box and just including them with BOOST_REGEX_INCLUDE_DIR at the cmake call

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 7b791c94c00158b90a09bb27e0a7101343f6d251
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |3 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |24 sec  (00:00:24) |
|setup (thor) |11 |11 |0 |47 sec  (00:00:47) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |860 |860 |0 |209 sec  (00:03:29) |
|test (thor) |784 |784 |0 |710 sec  (00:11:50) |
|test (roxie) |935 |935 |0 |419 sec  (00:06:59) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   34 sec (00:00:34) |  257 sec (00:04:17) |    0 sec (00:00:00) |    3 sec (00:00:03) |   19 sec (00:00:19) | 1653 sec (00:27:33) |   21 sec (00:00:21) | 1987 sec (00:33:07) |

richardkchapman commented:
Turning boost regex off (and using std instead) changes the semantics of regular expressions (AFAIK std library does not support perl syntax) so isn't really a great idea for a "standard" build.

HPCCSmoketest commented:
Process of PR-12959, label: centos-6-workaround-7.6.0 starts now.
The reason of this test is: New commit.
Commit ID: ac53e965d01277fa3a7b1e455787da3341d48da5
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: ac53e965d01277fa3a7b1e455787da3341d48da5
Number of Error(s): 2
make[1]: *** [CMakeFiles/Makefile2:16349: rtl/eclrtl/CMakeFiles/eclrtl.dir/all] Error 2
make: *** [Makefile:163: all] Error 2

[Build error:](https://gist.github.com/879cfc64214e445fd06941062365366e)
Build log [Build log](https://gist.github.com/HPCCSmoketest/879cfc64214e445fd06941062365366e#file-result-19-09-11-17-16-20-log)

xwang2713 reviewed (commented):
Looks good. Just add install for Clienttools.  For CentOS 6 build it requires "-DCENTOS_6_BOOST=ON" for all related projects

xwang2713 commented:
@AttilaVamos did you set "-DCENTOS_6_BOOST=ON"?

AttilaVamos commented:
@xwang2713 Should I do that on CentOS 7?

Michael-Gardner commented:
@AttilaVamos It should only be set on CentOS 6

AttilaVamos commented:
@Michael-Gardner back to you.

Michael-Gardner commented:
@richardkchapman ready for merge

HPCCSmoketest commented:
Process of PR-12959, label: centos-6-workaround-7.6.0 starts now.
The reason of this test is: New commit.
Commit ID: a2c24672974c6cff9c95083e134703f90855833f
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: a2c24672974c6cff9c95083e134703f90855833f
Number of Error(s): 2
make[1]: *** [CMakeFiles/Makefile2:16349: rtl/eclrtl/CMakeFiles/eclrtl.dir/all] Error 2
make: *** [Makefile:163: all] Error 2

[Build error:](https://gist.github.com/8edf36895a8b0e700aa20fca0e7c4a3b)
Build log [Build log](https://gist.github.com/HPCCSmoketest/8edf36895a8b0e700aa20fca0e7c4a3b#file-result-19-09-18-09-58-22-log)

HPCCSmoketest commented:
Process of PR-12959, label: centos-6-workaround-7.6.0 starts now.
The reason of this test is: New commit.
Commit ID: 28b0f1504c45e9db03c061b8933a4eebcfb82c23
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 28b0f1504c45e9db03c061b8933a4eebcfb82c23
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |29 sec  (00:00:29) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |17 sec  (00:00:17) |
|test (hthor) |860 |860 |0 |195 sec  (00:03:15) |
|test (thor) |784 |784 |0 |703 sec  (00:11:43) |
|test (roxie) |935 |935 |0 |235 sec  (00:03:55) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   44 sec (00:00:44) |  256 sec (00:04:16) |    0 sec (00:00:00) |    2 sec (00:00:02) |   20 sec (00:00:20) | 1447 sec (00:24:07) |   20 sec (00:00:20) | 1789 sec (00:29:49) |

Michael-Gardner commented:
@richardkchapman ready for merge

richardkchapman commented on cmake_modules/commonSetup.cmake:
```diff
@@ -58,6 +58,7 @@ IF ("${COMMONSETUP_DONE}" STREQUAL "")
   option(USE_OPENLDAP "Enable OpenLDAP support (requires OpenLDAP)" ON)
   option(USE_ICU "Enable unicode support (requires ICU)" ON)
   option(USE_BOOST_REGEX "Configure use of boost regex" ON)
+  option(CENTOS_6_BOOST "Supply regex library on CentOS 6" OFF)
```
Is there a reason we can't detect when this needs to be set?

Michael-Gardner replied:
I'll port over some logic to detect os and version so we can start automating this kind of thing instead of expecting the developer to set the flags.  https://track.hpccsystems.com/browse/HPCC-22859

mckellyln commented:
NOTE: see JIRA issue https://track.hpccsystems.com/browse/HPCC-23471
